tests/test_modeling_tf_common.py

```diff
@@ -2190,6 +2190,39 @@ def test_checkpoint_sharding_local(self):
                 for p1, p2 in zip(model.weights, new_model.weights):
                     self.assertTrue(np.allclose(p1.numpy(), p2.numpy()))
 
+    def test_save_pretrained_signatures(self):
+        model = TFBertModel.from_pretrained("hf-internal-testing/tiny-random-bert")
+
+        # Short custom TF signature function.
+        # `input_signature` is specific to BERT.
+        @tf.function(
+            input_signature=[
+                [
+                    tf.TensorSpec([None, None], tf.int32, name="input_ids"),
+                    tf.TensorSpec([None, None], tf.int32, name="token_type_ids"),
+                    tf.TensorSpec([None, None], tf.int32, name="attention_mask"),
+                ]
+            ]
+        )
+        def serving_fn(input):
+            return model(input)
+
+        with tempfile.TemporaryDirectory() as tmp_dir:
+            # Using default signature (default behavior)
+            model.save_pretrained(tmp_dir, saved_model=True, signatures=None)
+            model_loaded = tf.keras.models.load_model(f"{tmp_dir}/saved_model/1")
+            self.assertTrue(len(list(model_loaded.signatures.keys())) > 0)
+
+            # Providing custom signature function
+            model.save_pretrained(tmp_dir, saved_model=True, signatures=serving_fn)
+            model_loaded = tf.keras.models.load_model(f"{tmp_dir}/saved_model/1")
+            self.assertTrue(len(list(model_loaded.signatures.keys())) > 0)
+
+            # Providing custom signature function (dict input)
+            model.save_pretrained(tmp_dir, saved_model=True, signatures={"serving_default": serving_fn})
+            model_loaded = tf.keras.models.load_model(f"{tmp_dir}/saved_model/1")
+            self.assertTrue(len(list(model_loaded.signatures.keys())) > 0)
```

Comment:
There are a few issues with the current test :p 
1. The test asserts that there is a signature, but at some point in the future, we might remove the logic related to the argument you added in this PR (`signatures`) and the tests would still pass because the default signature would be loaded. We should test that we get a different signature when we pass a signature. For instance, the new signature could have two inputs (`input_ids` and `attention_mask`), and we could test the signature lengths are `3`, `2`, and `2`, following your test order.
2. I'm unsure whether saving a model into the same folder will store the model in a subfolder ending in `/1`, or if there might be other conflicting behavior. To be safe, add a `with tempfile.TemporaryDirectory() as tmp_dir:` context before each test, to ensure it stores the model on a new folder.

Reply:
Hey there @gante , thanks for the suggestions, I made a few updates to address them, let me know if it looks better now.

Regarding your points:
>1. The test asserts that there is a signature, but at some point in the future, we might remove the logic related to the argument you added in this PR (signatures) and the tests would still pass because the default signature would be loaded. We should test that we get a different signature when we pass a signature. For instance, the new signature could have two inputs (input_ids and attention_mask), and we could test the signature lengths are 3, 2, and 2, following your test order.
This is a great point, as you can see I updated the tests to check for the custom names of the signatures, I am not aware of a way to check the input names of the signature as you suggested, I can only check the signature names if you know how to do it let me know and I can update the tests.

>2. I'm unsure whether saving a model into the same folder will store the model in a subfolder ending in /1, or if there might be other conflicting behavior. To be safe, add a with tempfile.TemporaryDirectory() as tmp_dir: context before each test, to ensure it stores the model on a new folder.
Actually, the subsequent model savings will subscribe the previous ones, but your suggestion is pretty good since it makes the tests cleaner, thanks!